### PR TITLE
Add labels and scoped roles to the grammar and parsers

### PR DESCRIPTION
**Pull request by flyingsilverfin**
## What is the goal of this PR?
To being changes decided in #107 , we modify the grammar to have `type_scoped` as well as the default `type` that is unscoped. To make this make sense at the object level, we create a new object called `Label` that we will soon remove from `core` and `client-java` that can be scoped and unscoped types - Labels are now used throughout Graql to represent types. Where possible, we provide helper methods that consume `String`, creating the required `Label` objects internally and passing the Labels as required.

## What are the changes implemented in this PR?
* Create `Label`, a class holding the combination of scope + name. `scope` may be null
* The parser creates `Label` classes instead of strings from `label_ref`, `label_scoped`, and `label` rules
* `Compute` targets such as `of` and `in` require `Label` objects under the hood, though we provide convenience helpers that take strings and wrap them as `Label` where possible

### CODE REVIEW

File: java/statement/Label.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+    /**
+     * @param name The string which potentially represents a Type
+     * @return The matching Type Label
+     */
+    @CheckReturnValue
+    public static Label of(String name, @Nullable String scope) {
+        return new Label(name, scope);
+    }
```
Comment (flyingsilverfin) [inline java/statement/Label.java:55]: We could include another `.of(String name)` that doesn't require passing a null scope sometimes

File: java/statement/Label.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+    public String name() {
+        return name;
+    }
```
Comment (flyingsilverfin) [inline java/statement/Label.java:59]: Also question here whether this name() method should return the full name WITH scope, or the default non-scoped name

File: java/statement/Label.java (hunk 1 above)
Comment (haikalpribadi) [reply java/statement/Label.java:55]: Yes, add `.of(String name)`. We should only use `@Nullable` when we *really* have to.

File: java/statement/Label.java (hunk 2 above)
Comment (haikalpribadi) [reply java/statement/Label.java:59]: `name()` can return just the unscoped label, and we can add `scopedName()` and return WITH scope.

File: java/Graql.java [hunk 3]
```diff
@@ -247,15 +248,20 @@ private static Statement hiddenVar() {
     }
 
     @CheckReturnValue
-    public static StatementType type(Graql.Token.Type type) {
-        return type(type.toString());
+    public static StatementType type(Label label) {
+        return hiddenVar().type(label);
```
Comment (haikalpribadi) [inline java/Graql.java:252]: Why are we replacing `.type(Graql.Token.Type type)` with `.type(Label label)`, Josh? This was supposed to be used strictly for "native" types, right? Also, there are already 2 methods taking in labels as Strings below; they should be enough, right? Do we need one that strictly takes in an argument of `Class<Label>`? The user will then have to manually construct a `Label` class instance, to pass in, right?

File: java/property/TypeProperty.java [hunk 4]
```diff
@@ -27,17 +28,17 @@
...
             throw new NullPointerException("Null label");
         }
-        this.name = name;
+        this.label = label;
     }
 
-    public String name() {
-        return name;
+    public Label label() {
+        return label;
```
Comment (haikalpribadi) [inline java/property/TypeProperty.java:41]: Have we considered not having to implement a `Label` class, and just have `TypeProperty` to have `label` and `scope`, where both are represented as Strings?

File: java/statement/builder/StatementTypeBuilder.java [hunk 5]
```diff
@@ -46,72 +47,67 @@ default StatementType type(Graql.Token.Type type) {
...
         return type(AbstractProperty.get());
     }
 
     default StatementType sub(Graql.Token.Type type) {
-        return sub(type.toString());
+        return sub(type.toString(), null);
+    }
+
+    @CheckReturnValue
+    default StatementType sub(String type, @Nullable String scope) {
```
Comment (haikalpribadi) [inline java/statement/builder/StatementTypeBuilder.java:75]: I think we don't need the annotation `@Nullable` here?

File: java/statement/builder/StatementTypeBuilder.java [hunk 6]
```diff
@@ -46,72 +47,67 @@ default StatementType type(Graql.Token.Type type) {
...
-     */
     @CheckReturnValue
     default StatementType sub(Statement type) {
         return sub(new SubProperty(type));
     }
 
+    @CheckReturnValue
+    default StatementType sub(SubProperty property) {
+        return type(property);
+    }
```
Comment (haikalpribadi) [inline java/statement/builder/StatementTypeBuilder.java:92]: Why do we suddenly need this? 🤔

File: java/statement/builder/StatementTypeBuilder.java [hunk 7]
```diff
@@ -46,72 +47,67 @@ default StatementType type(Graql.Token.Type type) {
...
     }
 
-    /**
-     * @param type a concept type id that this variable must be a kind of, without looking at parent types
-     * @return this
-     */
     @CheckReturnValue
-    default StatementType subX(String type) {
-        return subX(Graql.type(type));
+    default StatementType subX(String type, @Nullable String scope) {
```
Comment (haikalpribadi) [inline java/statement/builder/StatementTypeBuilder.java:99]: This also doesn't need `@Nullable`. The user of this class needs to decide which method to call if they have or not have `scope`.

File: java/statement/builder/StatementTypeBuilder.java [hunk 8]
```diff
@@ -192,7 +191,8 @@ default StatementType relates(Statement type) {
      */
     @CheckReturnValue
     default StatementType relates(String roleType, @Nullable String superRoleType) {
```
Comment (haikalpribadi) [inline java/statement/builder/StatementTypeBuilder.java:193]: Should we also overload this?

File: java/property/TypeProperty.java (hunk 4 above)
Comment (flyingsilverfin) [reply java/property/TypeProperty.java:41]: yeah we need the class for the auto-generated parser, unless we want the auto-generated parser to return a Pair sometimes!

File: java/statement/builder/StatementTypeBuilder.java (hunk 5 above)
Comment (flyingsilverfin) [reply java/statement/builder/StatementTypeBuilder.java:75]: right

File: java/statement/builder/StatementTypeBuilder.java (hunk 8 above)
Comment (flyingsilverfin) [reply java/statement/builder/StatementTypeBuilder.java:193]: yeah lets stick to that style, I had not overloaded it to force me to update all the users properly

Comment (flyingsilverfin) [reply java/statement/builder/StatementTypeBuilder.java:193]: lets unify the other methods too

File: java/statement/builder/StatementTypeBuilder.java (hunk 6 above)
Comment (flyingsilverfin) [reply java/statement/builder/StatementTypeBuilder.java:92]: it existed before - just rearranged some of the methods

File: java/property/TypeProperty.java (hunk 4 above)
Comment (haikalpribadi) [reply java/property/TypeProperty.java:41]: Right

File: java/statement/builder/StatementTypeBuilder.java (hunk 6 above)
Comment (haikalpribadi) [reply java/statement/builder/StatementTypeBuilder.java:92]: ok

File: java/statement/builder/StatementTypeBuilder.java (hunk 8 above)
Comment (haikalpribadi) [reply java/statement/builder/StatementTypeBuilder.java:193]: and by 'unify' you mean overload other methods too, right? and remove `@Nullable` annotation?

File: java/statement/Label.java (hunk 2 above)
Comment (flyingsilverfin) [reply java/statement/Label.java:69]: yep that's what I've got!

**Pull request merged**